packages/tokamak/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol

```diff
@@ -54,4 +79,141 @@ contract L1CrossDomainMessenger is CrossDomainMessenger, ISemver {
     function _isUnsafeTarget(address _target) internal view override returns (bool) {
         return _target == address(this) || _target == address(PORTAL);
     }
+
+    /// @notice Sends a deposit ton message to some target address on the other chain. Note that if the call
+    ///         always reverts, then the message will be unrelayable, and any ETH sent will be
+    ///         permanently locked. The same will occur if the target on the other chain is
+    ///         considered unsafe (see the _isUnsafeTarget() function).
+    /// @param _target      Target contract or wallet address.
+    /// @param _amount      Amount of deposit TON.
+    /// @param _message     Message to trigger the target address with.
+    /// @param _minGasLimit Minimum gas limit that the message can be executed with.
+    function sendTONMessage(address _target, uint256 _amount, bytes calldata _message, uint32 _minGasLimit) external {
+        // Triggers a message to the other messenger. Note that the amount of gas provided to the
+        // message is the amount of gas requested by the user PLUS the base gas value. We want to
+        // guarantee the property that the call to the target contract will always have at least
+        // the minimum gas limit specified by the user.
+        _sendMessage(
+            OTHER_MESSENGER,
+            baseGas(_message, _minGasLimit),
+            _amount,
+            abi.encodeWithSelector(
+                this.relayMessage.selector, messageNonce(), msg.sender, _target, _amount, _minGasLimit, _message
+            )
+        );
+
+        emit SentMessage(_target, msg.sender, _message, messageNonce(), _minGasLimit);
+        emit SentMessageExtension1(msg.sender, _amount);
+
+        unchecked {
+            ++msgNonce;
+        }
+    }
+
+    /// @notice Relays a message that was sent by the other CrossDomainMessenger contract. Can only
+    ///         be executed via cross-chain call from the other messenger OR if the message was
+    ///         already received once and is currently being replayed.
+    /// @param _nonce       Nonce of the message being relayed.
+    /// @param _sender      Address of the user who sent the message.
+    /// @param _target      Address that the message is targeted at.
+    /// @param _value       TON value to send with the message.
+    /// @param _minGasLimit Minimum amount of gas that the message can be executed with.
+    /// @param _message     Message to send to the target.
+    function relayMessage(
```

Comment:
Both `sendMessage` and `relayMessage` exist on `CrossDomainMessenger.sol`, why don't you modify them instead of making new functions on L1CrossDomainMessager? 😊

Reply:
I think you are right, move these functions into `L1CrossDomainMessager` to reduce the confusion about what is `msg.vaue` presents to.

In `L1CrossDomainMessager`, `msg.value` represents as ETH token, but in `L2CrossDomainMessager`, `msg.value` represents as TON token

Reply:
We have to prevent dep/with eth in legacy way. (L1: native ETH <-> L2: native ETH) If we make some function for TON, I think we need additional code. 🙇🏼‍♂️